proposals/ssss-distributed-actor-runtime.md

```diff
@@ -0,0 +1,1798 @@
+    }
+    
+    return subTypes
+  }
+  
+  /// Attempt to decode the next argument from the underlying buffers into pre-allocated storage
+  /// pointed at by 'pointer'.
+  ///
+  /// This method should throw if it has no more arguments available, if decoding the argument failed,
+  /// or, optionally, if the argument type we're trying to decode does not match the stored type.
+  ///
+  /// The result of the decoding operation must be stored into the provided 'pointer' rather than
+  /// returning a value. This pattern allows the runtime to use a heavily optimized, pre-allocated
+  /// buffer for all the arguments and their expected types. The 'pointer' passed here is a pointer
+  /// to a "slot" in that pre-allocated buffer. That buffer will then be passed to a thunk that
+  /// performs the actual distributed (local) instance method invocation.
+  mutating func decodeNext<Argument: SerializationRequirement>() throws {
+    try nextDataLength = try bytes.readInt()
+    let nextData = try bytes.readData(bytes: nextDataLength)
+    // since we are guaranteed the values are Codable, so we can just invoke it:
+    return try system.decoder.decode(as: Argument.self, from: bytes)
+  }
+
+  mutating func decodeErrorType() throws -> Any.Type? { 
+    let length = try self.bytes.readInt() // read the length of the type
+    guard length > 0 {
+      return nil // we don't always transmit it, 0 length means "none"
+    }
+    let typeName = try self.bytes.readString(length: length)
+    return try self.system.summonType(byName: typeName)
+  }
+
+  mutating func decodeReturnType() throws -> Any.Type? { 
+    let length = try self.bytes.readInt() // read the length of the type
+    guard length > 0 {
+      return nil // we don't always transmit it, 0 length means "none"
+    }
+    let typeName = try self.bytes.readString(length: length)
+    return try self.system.summonType(byName: typeName)
+  }
+}
+```
+
+The general idea here is that the `Invocation` is *lazy* in its decoding and just stores the remaining bytes of the envelope. All we need to do for now is to implement the Invocation in such way that it expects the decoding methods be invoked in the following order (which is the same as the order on the sending side):
+
+- 0...1 invocation of `decodeGenericArguments`,
+- 0...n invocations of `decoder.decodeNextArgument<Argument>`,
+- 0...1 invocations of `decodeReturnType`,
+- 0...1 invocations of `decodeErrorType`.
+
+Decoding arguments is the most interesting here. This is another case where the compiler and Swift runtime enables us to implement things more easily. Since the `Argument` generic type of the `decodeNextArgument` is ensured to conform to the `SerializationRequirement`, actor system implementations can rely on this fact and have a simpler time implementing the decoding steps. For example, with `Codable` the decoding steps becomes a rather simple task of invoking the usual `Decoder` APIs.
```

Comment:
```suggestion
Decoding arguments is the most interesting here. This is another case where the compiler and Swift runtime enable us to implement things more easily. Since the `Argument` generic type of the `decodeNextArgument` is ensured to conform to the `SerializationRequirement`, actor system implementations can rely on this fact and have a simpler time implementing the decoding steps. For example, with `Codable` the decoding steps becomes a rather simple task of invoking the usual `Decoder` APIs.
```